# Analysis of Goolge Merchant Store Analytics data

Google Analytics had conducted a competition on Kaggle that aimed to predict revenue users generate based on their behviour captured by Google Analytics themselves. Though the competition suffered from feature leakage and the fact that participants found the complete dataset online; when using the data provided , several key concepts of data science and ML can be put to use here.

# Problem statement and nature of data
Google wanted to predict transaction a particular visitor might generate when they visit the google merchant website. A dataset was provided in form of a flatfile with embedded JSON which was to be used to find the best trends.

### Proceeding with imports to provide a view of the data

In [37]:
import json
from pprint import pprint
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import numpy as np
import datetime as dt
import urllib
import  pickle
import sys, os
import re
pd.set_option('display.max_rows', 5000)
pd.set_option('display.max_columns', 5000)
pd.set_option('display.width', 10000)
pd.options.display.max_colwidth = 1000
## Above options to provide a decent view of the data in tablular form

In [7]:
#trainRaw = pd.read_csv(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\train.csv' , dtype = {'fullVisitorId':'|S'})
## Column fullVisitorId is used as String as it has leading zeros in some instances

In [ ]:
trainRaw.head(1)

The first row from the flat file shows the arrnagement of the JSON structure viz. Stored as columns

channelGrouping      date                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           device        fullVisitorId                                                                                                                                                                                                                                                                                                                                                       geoNetwork                       sessionId  socialEngagementType                                                                            totals                                                                                                                                                                 trafficSource     visitId  visitNumber  visitStartTime  visitStartTime_x  diff  visits  hits  pageviews  bototalsunces  newVisits  transactionRevenue browser                 browserVersion                    browserSize operatingSystem         operatingSystemVersion  isMobile           mobileDeviceBranding              mobileDeviceModel            mobileInputSelector               mobileDeviceInfo      mobileDeviceMarketingName                   flashVersion                       language                   screenColors               screenResolution deviceCategory   campaign  source   medium         keyword isTrueDirect referralPath adContent campaignCode continent  subContinent country region      metro   city                         cityId networkDomain                       latitude                      longitude                networkLocation
0  Organic Search  20160902  {"browser": "Chrome", "browserVersion": "not available in demo dataset", "browserSize": "not available in demo dataset", "operatingSystem": "Windows", "operatingSystemVersion": "not available in demo dataset", "isMobile": false, "mobileDeviceBranding": "not available in demo dataset", "mobileDeviceModel": "not available in demo dataset", "mobileInputSelector": "not available in demo dataset", "mobileDeviceInfo": "not available in demo dataset", "mobileDeviceMarketingName": "not available in demo dataset", "flashVersion": "not available in demo dataset", "language": "not available in demo dataset", "screenColors": "not available in demo dataset", "screenResolution": "not available in demo dataset", "deviceCategory": "desktop"}  1131660440785968503  {"continent": "Asia", "subContinent": "Western Asia", "country": "Turkey", "region": "Izmir", "metro": "(not set)", "city": "Izmir", "cityId": "not available in demo dataset", "networkDomain": "ttnet.com.tr", "latitude": "not available in demo dataset", "longitude": "not available in demo dataset", "networkLocation": "not available in demo dataset"}  1131660440785968503_1472830385  Not Socially Engaged  {"visits": "1", "hits": "1", "pageviews": "1", "bounces": "1", "newVisits": "1"}  {"campaign": "(not set)", "source": "google", "medium": "organic", "keyword": "(not provided)", "adwordsClickInfo": {"criteriaParameters": "not available in demo dataset"}}  1472830385            1      1472830385        1472830385     0       1     1          1              0          1                   0  Chrome  not available in demo dataset  not available in demo dataset         Windows  not available in demo dataset     Fa

The data was converted to a flat strucutre using a JSON parsing code : Courtesy one of the participants who shared
# This script read the csvs, delete useless columns, set datatypes
# and save dataframes as pickle files for faster loading

import os
import numpy as np
import pandas as pd
from pandas.io.json import json_normalize
import json
import time

pd.options.display.max_columns = 999

def load_df(file_name, nrows=None):
    """ Read csv and convert json columns. Author: Julián Peller. """
    
    JSON_COLUMNS = ['device', 'geoNetwork', 'totals', 'trafficSource']
    df = pd.read_csv(file_name,
                     converters={column: json.loads for column in JSON_COLUMNS}, 
                     dtype={'fullVisitorId': 'str'}, nrows=nrows)
    
    for column in JSON_COLUMNS:
        column_as_df = json_normalize(df[column])
        column_as_df.columns = [f"{column}_{subcolumn}" for subcolumn in column_as_df.columns]
        df = df.drop(column, axis=1).merge(column_as_df, right_index=True, left_index=True)
    return df

    # Read csv
timer = time.time()
#train = load_df(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\train.csv' )
#test = load_df(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\test.csv')

The data that was previously pickled is loaded into pandas DF again
with open(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\pickles\train_clean1.pkl','rb') as filePointer:
    train = pickle.load(filePointer)


In [14]:
train.head(1)

channelGrouping      date        fullVisitorId                       sessionId  socialEngagementType     visitId  visitNumber  visitStartTime device_browser             device_browserSize          device_browserVersion device_deviceCategory            device_flashVersion  device_isMobile                device_language    device_mobileDeviceBranding        device_mobileDeviceInfo device_mobileDeviceMarketingName       device_mobileDeviceModel     device_mobileInputSelector device_operatingSystem  device_operatingSystemVersion            device_screenColors        device_screenResolution geoNetwork_city              geoNetwork_cityId geoNetwork_continent geoNetwork_country            geoNetwork_latitude           geoNetwork_longitude geoNetwork_metro geoNetwork_networkDomain     geoNetwork_networkLocation geoNetwork_region geoNetwork_subContinent totals_bounces totals_hits totals_newVisits totals_pageviews totals_transactionRevenue totals_visits trafficSource_adContent trafficSource_adwordsClickInfo.adNetworkType trafficSource_adwordsClickInfo.criteriaParameters trafficSource_adwordsClickInfo.gclId trafficSource_adwordsClickInfo.isVideoAd trafficSource_adwordsClickInfo.page trafficSource_adwordsClickInfo.slot trafficSource_campaign trafficSource_campaignCode trafficSource_isTrueDirect trafficSource_keyword trafficSource_medium trafficSource_referralPath trafficSource_source
0  Organic Search  20160902  1131660440785968503  1131660440785968503_1472830385  Not Socially Engaged  1472830385            1      1472830385         Chrome  not available in demo dataset  not available in demo dataset               desktop  not available in demo dataset            False  not available in demo dataset  not available in demo dataset  not available in demo dataset    not available in demo dataset  not available in demo dataset  not available in demo dataset                Windows  not available in demo dataset  not available in demo dataset  not available in demo dataset           Izmir  not available in demo dataset                 Asia             Turkey  not available in demo dataset  not available in demo dataset        (not set)             ttnet.com.tr  not available in demo dataset             Izmir            Western Asia              1           1                1                1                       NaN             1                     NaN                                          NaN                     not available in demo dataset                                  NaN                                      NaN                                 NaN                                 NaN              (not set)                        NaN                        NaN        (not provided)              organic                        NaN               google

# Important feature discovery
It was observed that apart from the usual features , the timezone of the request was extremely helpful.Argument being, many individuals would prefer to shop during a convinient time in the day and it might lead to a purchase.
The time in the dataset was the time of the IP destination and not the time of the origin city . 

### We hence took some commercially available geolocation and timezone apis and created a file that contained the timezone and time offset of the cities that have appeared in the data. This timezone info was offset to get the local time when the request when the visit was made. Moreover using cities we can also get an idea of the probably currency of transaction and adjust the transaction amounts whose units are kept hidden

In [ ]:
locations = pd.read_csv(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\export\cityrequestfile.txt')
locations.head(2)
import requests
file = open(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\export\latlong.csv','a') 
for locnames in locations.location:
    r = requests.get(r'https://nominatim.openstreetmap.org/search?q='+locnames+'&format=json&addressdetails=0')
    t=json.loads(r.content)
    for i in range(1):
        print(locnames+','+t[i]['lat']+','+t[i]['lon'])
        file.writelines(locnames+','+t[i]['lat']+','+t[i]['lon']+'\n')
file.close()


### Shwon below is a sample of the responses recived from the request. We used google geolocation and timezone apis

#### Examples
Tajikistan,38.6281733,70.8156541<br>
Tanzania,-6.5247123,35.7878438<br>
Thailand,14.8971921,100.83273<br>
Hua Hin,12.5678487,99.9575368<br>
Nakhon Pathom,13.8918425,100.0165659<br>
Timor-Leste,-8.5151979,125.8375756<br>
Togo,8.7800265,1.0199765<br>
Trinidad & Tobago,37.169397,-104.5005411<br>
Tunisia,33.8439408,9.400138<br>
Turkey,38.9597594,34.9249653<br>



In [24]:
locationtimezoneoffsets = pd.read_csv(r'C:\Users\This PC\Documents\NewTechnologies\MachineLearning\data\GoogleAnalytics\all\export\resp1.txt' , sep='~')
locationtimezoneoffsets = locationtimezoneoffsets.rename({'location':'geoNetwork_city'} , axis =1)
locationtimezoneoffsets.head(1)


geoNetwork_city  dst   timezoneId  offset              timezoneName
0          Manila    0  Asia/Manila   28800  Philippine Standard Time

The next steps being performed are to clean some remaining data. , Default values that do not exist and impute some non numerical values into numeric features
Most of the code is self explanatory. The fetures being used finally will be described in the sections further below


In [25]:
## Renaming a column before the Join
train2=train.copy()
train = train.merge(locationtimezoneoffsets , left_on =['geoNetwork_city'] ,right_on=['geoNetwork_city'] , how = 'left')
locationtimezoneoffsets = locationtimezoneoffsets.rename({'geoNetwork_city':'geoNetwork_country'} , axis =1) 
#Join is peformed between timezone file and the train data
train2 = train2.merge(locationtimezoneoffsets , left_on =['geoNetwork_country'] ,right_on=['geoNetwork_country'] , how = 'left')


In [27]:
train.head(1)

channelGrouping      date        fullVisitorId                       sessionId  socialEngagementType     visitId  visitNumber  visitStartTime device_browser             device_browserSize          device_browserVersion device_deviceCategory            device_flashVersion  device_isMobile                device_language    device_mobileDeviceBranding        device_mobileDeviceInfo device_mobileDeviceMarketingName       device_mobileDeviceModel     device_mobileInputSelector device_operatingSystem  device_operatingSystemVersion            device_screenColors        device_screenResolution geoNetwork_city              geoNetwork_cityId geoNetwork_continent geoNetwork_country            geoNetwork_latitude           geoNetwork_longitude geoNetwork_metro geoNetwork_networkDomain     geoNetwork_networkLocation geoNetwork_region geoNetwork_subContinent totals_bounces totals_hits totals_newVisits totals_pageviews totals_transactionRevenue totals_visits trafficSource_adContent trafficSource_adwordsClickInfo.adNetworkType trafficSource_adwordsClickInfo.criteriaParameters trafficSource_adwordsClickInfo.gclId trafficSource_adwordsClickInfo.isVideoAd trafficSource_adwordsClickInfo.page trafficSource_adwordsClickInfo.slot trafficSource_campaign trafficSource_campaignCode trafficSource_isTrueDirect trafficSource_keyword trafficSource_medium trafficSource_referralPath trafficSource_source  dst       timezoneId  offset                    timezoneName
0  Organic Search  20160902  1131660440785968503  1131660440785968503_1472830385  Not Socially Engaged  1472830385            1      1472830385         Chrome  not available in demo dataset  not available in demo dataset               desktop  not available in demo dataset            False  not available in demo dataset  not available in demo dataset  not available in demo dataset    not available in demo dataset  not available in demo dataset  not available in demo dataset                Windows  not available in demo dataset  not available in demo dataset  not available in demo dataset           Izmir  not available in demo dataset                 Asia             Turkey  not available in demo dataset  not available in demo dataset        (not set)             ttnet.com.tr  not available in demo dataset             Izmir            Western Asia              1           1                1                1                       NaN             1                     NaN                                          NaN                     not available in demo dataset                                  NaN                                      NaN                                 NaN                                 NaN              (not set)                        NaN                        NaN        (not provided)              organic                        NaN               google  0.0  Europe/Istanbul  7200.0  Eastern European Standard Time

In [28]:
train_city = train.loc[~train.offset.isnull()]
train_ctry = train2.loc[~(train2.offset.isnull()) & ~(train2.sessionId.isin(train_city.sessionId))]
train_tz = train_ctry.append(train_city, ignore_index=True)
train_tz = train_tz.assign(unixtimeinzone = lambda x : x['offset'].astype(np.int64) + x['visitStartTime'].astype(np.int64) )
train_tz = train_tz.assign (month =  lambda x : pd.to_datetime(x['unixtimeinzone'], unit='s').dt.month, hour =  lambda x : pd.to_datetime(x['unixtimeinzone'], unit='s').dt.hour  ,day =  lambda x : pd.to_datetime(x['unixtimeinzone'], unit='s').dt.day )
## The offsets are added to the local time is derived


In [30]:
train_tz.head(2)

channelGrouping      date       fullVisitorId                      sessionId  socialEngagementType     visitId  visitNumber  visitStartTime device_browser             device_browserSize          device_browserVersion device_deviceCategory            device_flashVersion  device_isMobile                device_language    device_mobileDeviceBranding        device_mobileDeviceInfo device_mobileDeviceMarketingName       device_mobileDeviceModel     device_mobileInputSelector device_operatingSystem  device_operatingSystemVersion            device_screenColors        device_screenResolution                geoNetwork_city              geoNetwork_cityId geoNetwork_continent geoNetwork_country            geoNetwork_latitude           geoNetwork_longitude               geoNetwork_metro geoNetwork_networkDomain     geoNetwork_networkLocation              geoNetwork_region geoNetwork_subContinent totals_bounces totals_hits totals_newVisits totals_pageviews totals_transactionRevenue totals_visits trafficSource_adContent trafficSource_adwordsClickInfo.adNetworkType trafficSource_adwordsClickInfo.criteriaParameters trafficSource_adwordsClickInfo.gclId trafficSource_adwordsClickInfo.isVideoAd trafficSource_adwordsClickInfo.page trafficSource_adwordsClickInfo.slot trafficSource_campaign trafficSource_campaignCode trafficSource_isTrueDirect trafficSource_keyword trafficSource_medium trafficSource_referralPath trafficSource_source  dst        timezoneId   offset                      timezoneName  unixtimeinzone  month  hour  day
0  Organic Search  20160902  377306020877927890  377306020877927890_1472880147  Not Socially Engaged  1472880147            1      1472880147        Firefox  not available in demo dataset  not available in demo dataset               desktop  not available in demo dataset            False  not available in demo dataset  not available in demo dataset  not available in demo dataset    not available in demo dataset  not available in demo dataset  not available in demo dataset              Macintosh  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset              Oceania          Australia  not available in demo dataset  not available in demo dataset  not available in demo dataset              dodo.net.au  not available in demo dataset  not available in demo dataset             Australasia              1           1                1                1                       NaN             1                     NaN                                          NaN                     not available in demo dataset                                  NaN                                      NaN                                 NaN                                 NaN              (not set)                        NaN                        NaN        (not provided)              organic                        NaN               google  0.0  Australia/Darwin  34200.0  Australian Central Standard Time      1472914347      9    14    3
1  Organic Search  20160902   27294437909732085   27294437909732085_1472822600  Not Socially Engaged  1472822600            2      1472822600         Chrome  not available in demo dataset  not available in demo dataset                mobile  not available in demo dataset             True  not available in demo dataset  not available in demo dataset  not available in demo dataset    not available in demo dataset  not available in demo dataset  not available in demo dataset                Android  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset               Europe     United Kingdom  not available in demo dataset  not available in demo dataset  not available in demo dataset          unknown.unknown  not available in demo dataset  not available in demo dataset         Northern Europe              1           1   

### The new columns derived are below

timezoneId 	offset 	timezoneName 	unixtimeinzone 	month 	hour 	day<br>
Australian Central Standard Time 	1472914347 	9 	14 	3<br>
Europe/London 	0.0 	Greenwich Mean Time 	1472822600 	9 	13 	2<br>

These columns will be used a features .They collectiviely show the local time down to the hour


 Some data columns are cleaned and defaulted to preset values.
 
 Most the of the transactions had occured in either Google Americas or Google Rest of the world store
 Rest of world stores used GBP (Pound) sterling as currency in transation amount. hence we are bumping up transaction
 amount by approx 1.32 times for GBP transactions (We did not have this particular info , and used research and web site)
 to devise the same

In [33]:
train_tz['visitNumber'] = train['visitNumber'].astype('int16')
train_tz['totals_hits'] = train['totals_hits'].astype('int16')
train_tz['totals_pageviews'] = train['totals_pageviews'].astype('float64')
train_tz['trafficSource_adwordsClickInfo.page'] = train['trafficSource_adwordsClickInfo.page'].astype('float64')
train_tz['totals_newVisits'] = train['totals_newVisits'].astype('float64')
train_tz['totals_bounces'] = train['totals_bounces'].astype('float64')
train_tz['totals_transactionRevenue'] = train['totals_transactionRevenue'].fillna(0).astype('int64')
train_tz['totals_transactionRevenue'] = train_tz['totals_transactionRevenue'].loc[~(train_tz.geoNetwork_country.str.upper() == 'UNITED STATES') & ~(train_tz.geoNetwork_country.str.upper() == 'CANADA')]*1.32


In [35]:
train = train_tz.copy()

In [38]:
def impute_source(x):
    if('googleplex' in x.lower()):
        return 'googlestore'
    elif(np.bool(re.match('^google',x.lower()))):
        return 'googlewebsite'
    elif(np.bool(re.match('.*google',x.lower()))):
        return 'googlesupport'
    elif('(direct)' in x.lower()):
        return 'direct'
    else:
        return 'nongoogle'

k=train.apply(lambda x : impute_source( (x['trafficSource_source'])), axis = 1 )
k = pd.DataFrame(k ,columns = ['sourcecat'])
train = pd.concat([train ,k] , axis =1)

Data observation showed traffic originated from google affiliated employee domains
Like GooglePlex ,google websites (google search engine) , google domain like gamail etc  and Direct visits to the store
These features are likely to affect the purchasing decision as well
Hence they are being imputed into categories.


The following features are finally chosen for the algorthim. These are are needed to be one hot encoded for consumption
into a lightgbm tree boost algorithm
##### channelGrouping
##### geoNetwork_continent 
#### trafficSource_medium
#### sourcecat

The function below will one hot encode the said columns


In [39]:
def dataOneHotEncoder(df: pd.DataFrame , gbcols : list):
  for colname in gbcols:
      df = pd.concat([df,pd.get_dummies(df[colname], prefix=colname,dummy_na=True)],axis=1).drop([colname],axis=1)
      return df

In [57]:
#train = dataOneHotEncoder(train , [('channelGrouping'),('geoNetwork_continent') , ('trafficSource_medium') , ('sourcecat')])
#train = dataOneHotEncoder(train , [('trafficSource_medium')])

In [51]:
train.head(3)

date        fullVisitorId                       sessionId  socialEngagementType     visitId  visitNumber  visitStartTime device_browser             device_browserSize          device_browserVersion device_deviceCategory            device_flashVersion  device_isMobile                device_language    device_mobileDeviceBranding        device_mobileDeviceInfo device_mobileDeviceMarketingName       device_mobileDeviceModel     device_mobileInputSelector device_operatingSystem  device_operatingSystemVersion            device_screenColors        device_screenResolution                geoNetwork_city              geoNetwork_cityId geoNetwork_country            geoNetwork_latitude           geoNetwork_longitude               geoNetwork_metro geoNetwork_networkDomain     geoNetwork_networkLocation              geoNetwork_region geoNetwork_subContinent  totals_bounces  totals_hits  totals_newVisits  totals_pageviews  totals_transactionRevenue totals_visits trafficSource_adContent trafficSource_adwordsClickInfo.adNetworkType trafficSource_adwordsClickInfo.criteriaParameters trafficSource_adwordsClickInfo.gclId trafficSource_adwordsClickInfo.isVideoAd  trafficSource_adwordsClickInfo.page trafficSource_adwordsClickInfo.slot trafficSource_campaign trafficSource_campaignCode trafficSource_isTrueDirect trafficSource_keyword trafficSource_medium trafficSource_referralPath trafficSource_source  dst        timezoneId   offset                      timezoneName  unixtimeinzone  month  hour  day      sourcecat  channelGrouping_(Other)  channelGrouping_Affiliates  channelGrouping_Direct  channelGrouping_Display  channelGrouping_Organic Search  channelGrouping_Paid Search  channelGrouping_Referral  channelGrouping_Social  channelGrouping_nan  geoNetwork_continent_(not set)  geoNetwork_continent_Africa  geoNetwork_continent_Americas  geoNetwork_continent_Asia  geoNetwork_continent_Europe  geoNetwork_continent_Oceania  geoNetwork_continent_nan
0  20160902   377306020877927890   377306020877927890_1472880147  Not Socially Engaged  1472880147            1      1472880147        Firefox  not available in demo dataset  not available in demo dataset               desktop  not available in demo dataset            False  not available in demo dataset  not available in demo dataset  not available in demo dataset    not available in demo dataset  not available in demo dataset  not available in demo dataset              Macintosh  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset          Australia  not available in demo dataset  not available in demo dataset  not available in demo dataset              dodo.net.au  not available in demo dataset  not available in demo dataset             Australasia             1.0            1               1.0               1.0                        0.0             1                     NaN                                          NaN                     not available in demo dataset                                  NaN                                      NaN                                  NaN                                 NaN              (not set)                        NaN                        NaN        (not provided)              organic                        NaN               google  0.0  Australia/Darwin  34200.0  Australian Central Standard Time      1472914347      9    14    3  googlewebsite                        0                           0                       0                        0                               1                            0                         0                       0                    0                               0                            0                              0                          0                            0                             1                         0
1  20160902    27294437909732085    27294437909732085_1472822600  Not Socially Engage

Data post encoding looks like as above . The classes below are fucntions for cross validation and data splits


In [43]:
from sklearn.metrics import mean_squared_error
import gc
import time
from pandas.core.common import SettingWithCopyWarning
import warnings
import lightgbm as lgb
from sklearn.model_selection import GroupKFold
warnings.simplefilter('error', SettingWithCopyWarning)
gc.enable()
def get_folds(df=None, n_splits=5):
    """Returns dataframe indices corresponding to Visitors Group KFold"""
    # Get sorted unique visitors
    unique_vis = np.array(sorted(df['fullVisitorId'].unique()))

    # Get folds
    folds = GroupKFold(n_splits=n_splits)
    fold_ids = []
    ids = np.arange(df.shape[0])
    for trn_vis, val_vis in folds.split(X=unique_vis, y=unique_vis, groups=unique_vis):
        fold_ids.append(
            [
                ids[df['fullVisitorId'].isin(unique_vis[trn_vis])],
                ids[df['fullVisitorId'].isin(unique_vis[val_vis])]
            ]
        )

    return fold_ids

In [44]:
y_reg = train['totals_transactionRevenue']
folds = get_folds(df=train, n_splits=5)
importances = pd.DataFrame()
oof_reg_preds = np.zeros(train.shape[0])

In [59]:
train.head(3)

date        fullVisitorId                       sessionId  socialEngagementType     visitId  visitNumber  visitStartTime device_browser             device_browserSize          device_browserVersion device_deviceCategory            device_flashVersion  device_isMobile                device_language    device_mobileDeviceBranding        device_mobileDeviceInfo device_mobileDeviceMarketingName       device_mobileDeviceModel     device_mobileInputSelector device_operatingSystem  device_operatingSystemVersion            device_screenColors        device_screenResolution                geoNetwork_city              geoNetwork_cityId geoNetwork_country            geoNetwork_latitude           geoNetwork_longitude               geoNetwork_metro geoNetwork_networkDomain     geoNetwork_networkLocation              geoNetwork_region geoNetwork_subContinent  totals_bounces  totals_hits  totals_newVisits  totals_pageviews  totals_transactionRevenue totals_visits trafficSource_adContent trafficSource_adwordsClickInfo.adNetworkType trafficSource_adwordsClickInfo.criteriaParameters trafficSource_adwordsClickInfo.gclId trafficSource_adwordsClickInfo.isVideoAd  trafficSource_adwordsClickInfo.page trafficSource_adwordsClickInfo.slot trafficSource_campaign trafficSource_campaignCode trafficSource_isTrueDirect trafficSource_keyword trafficSource_referralPath trafficSource_source  dst        timezoneId   offset                      timezoneName  unixtimeinzone  month  hour  day  channelGrouping_(Other)  channelGrouping_Affiliates  channelGrouping_Direct  channelGrouping_Display  channelGrouping_Organic Search  channelGrouping_Paid Search  channelGrouping_Referral  channelGrouping_Social  channelGrouping_nan  geoNetwork_continent_(not set)  geoNetwork_continent_Africa  geoNetwork_continent_Americas  geoNetwork_continent_Asia  geoNetwork_continent_Europe  geoNetwork_continent_Oceania  geoNetwork_continent_nan  sourcecat_direct  sourcecat_googlestore  sourcecat_googlesupport  sourcecat_googlewebsite  sourcecat_nongoogle  sourcecat_nan  trafficSource_medium_(none)  trafficSource_medium_(not set)  trafficSource_medium_affiliate  trafficSource_medium_cpc  trafficSource_medium_cpm  trafficSource_medium_organic  trafficSource_medium_referral  trafficSource_medium_nan
0  20160902   377306020877927890   377306020877927890_1472880147  Not Socially Engaged  1472880147            1      1472880147        Firefox  not available in demo dataset  not available in demo dataset               desktop  not available in demo dataset            False  not available in demo dataset  not available in demo dataset  not available in demo dataset    not available in demo dataset  not available in demo dataset  not available in demo dataset              Macintosh  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset  not available in demo dataset          Australia  not available in demo dataset  not available in demo dataset  not available in demo dataset              dodo.net.au  not available in demo dataset  not available in demo dataset             Australasia             1.0            1               1.0               1.0                        0.0             1                     NaN                                          NaN                     not available in demo dataset                                  NaN                                      NaN                                  NaN                                 NaN              (not set)                        NaN                        NaN        (not provided)                        NaN               google  0.0  Australia/Darwin  34200.0  Australian Central Standard Time      1472914347      9    14    3                        0                           0                       0                        0                               1                            0                         0                       0                    0   

In [58]:
## The features chosne for the algorithm
X= train[['totals_hits','totals_newVisits','totals_pageviews','month','hour','day','channelGrouping_(Other)','channelGrouping_Affiliates','channelGrouping_Direct','channelGrouping_Display','channelGrouping_Organic Search','channelGrouping_Paid Search','channelGrouping_Referral','channelGrouping_Social','channelGrouping_nan','geoNetwork_continent_(not set)','geoNetwork_continent_Africa','geoNetwork_continent_Americas','geoNetwork_continent_Asia','geoNetwork_continent_Europe','geoNetwork_continent_Oceania','geoNetwork_continent_nan','sourcecat_direct','sourcecat_googlestore','sourcecat_googlesupport','sourcecat_googlewebsite','sourcecat_nongoogle','sourcecat_nan','trafficSource_medium_(none)','trafficSource_medium_(not set)','trafficSource_medium_affiliate','trafficSource_medium_cpc','trafficSource_medium_cpm','trafficSource_medium_organic','trafficSource_medium_referral','trafficSource_medium_nan']]

In [61]:
train_features = X.columns
#train_features
y_reg = y_reg.fillna(0)
X = X.fillna(0)
X.head(1)

totals_hits  totals_newVisits  totals_pageviews  month  hour  day  channelGrouping_(Other)  channelGrouping_Affiliates  channelGrouping_Direct  channelGrouping_Display  channelGrouping_Organic Search  channelGrouping_Paid Search  channelGrouping_Referral  channelGrouping_Social  channelGrouping_nan  geoNetwork_continent_(not set)  geoNetwork_continent_Africa  geoNetwork_continent_Americas  geoNetwork_continent_Asia  geoNetwork_continent_Europe  geoNetwork_continent_Oceania  geoNetwork_continent_nan  sourcecat_direct  sourcecat_googlestore  sourcecat_googlesupport  sourcecat_googlewebsite  sourcecat_nongoogle  sourcecat_nan  trafficSource_medium_(none)  trafficSource_medium_(not set)  trafficSource_medium_affiliate  trafficSource_medium_cpc  trafficSource_medium_cpm  trafficSource_medium_organic  trafficSource_medium_referral  trafficSource_medium_nan
0            1               1.0               1.0      9    14    3                        0                           0                       0                        0                               1                            0                         0                       0                    0                               0                            0                              0                          0                            0                             1                         0                 0                      0                        0                        1                    0              0                            0                               0                               0                         0                         0                             1                              0                         0

In [ ]:
oof_reg_preds = np.zeros(train.shape[0])
##sub_reg_preds = np.zeros(test.shape[0])
for fold_, (trn_, val_) in enumerate(folds):
    trn_x, trn_y = train[train_features].iloc[trn_], y_reg.iloc[trn_]
    val_x, val_y = train[train_features].iloc[val_], y_reg.iloc[val_]
    
    reg = lgb.LGBMRegressor(
        num_leaves=31,
        learning_rate=0.03,
        n_estimators=1000,
        subsample=.9,
        colsample_bytree=.9,
        random_state=6
    )
    reg.fit(
        trn_x, np.log1p(trn_y),
        eval_set=[(val_x, np.log1p(val_y))],
        early_stopping_rounds=50,
        verbose=100,
       categorical_feature = ['month','hour','day','channelGrouping_(Other)','channelGrouping_Affiliates','channelGrouping_Direct','channelGrouping_Display','channelGrouping_Organic Search','channelGrouping_Paid Search','channelGrouping_Referral','channelGrouping_Social','channelGrouping_nan','geoNetwork_continent_(not set)','geoNetwork_continent_Africa','geoNetwork_continent_Americas','geoNetwork_continent_Asia','geoNetwork_continent_Europe','geoNetwork_continent_Oceania','geoNetwork_continent_nan','trafficSource_medium_(none)','trafficSource_medium_(not set)','trafficSource_medium_affiliate','trafficSource_medium_cpc','trafficSource_medium_cpm','trafficSource_medium_organic','trafficSource_medium_referral','trafficSource_medium_nan','sourcecat_direct','sourcecat_googlestore','sourcecat_googlesupport','sourcecat_googlewebsite','sourcecat_nongoogle','sourcecat_nan'],
        eval_metric='rmse'
    )
    imp_df = pd.DataFrame()
    imp_df['feature'] = train_features
    imp_df['gain'] = reg.booster_.feature_importance(importance_type='gain')
    
    imp_df['fold'] = fold_ + 1
    importances = pd.concat(([importances, imp_df]), axis=0)
    
    oof_reg_preds[val_] = reg.predict(val_x, num_iteration=reg.best_iteration_)
    oof_reg_preds[oof_reg_preds < 0] = 0
   # _preds = reg.predict(test[train_features], num_iteration=reg.best_iteration_)
   # _preds[_preds < 0] = 0
    #sub_reg_preds += np.expm1(_preds) / len(folds)
    
mean_squared_error(np.log1p(y_reg), oof_reg_preds) ** .5

We found the least deviation observed was
Training until validation scores don't improve for 50 rounds.<br>
[100]	valid_0's rmse: 1.34795	valid_0's l2: 1.81696<br>
[200]	valid_0's rmse: 1.34569	valid_0's l2: 1.81089<br>
Early stopping, best iteration is:<br>
[229]	valid_0's rmse: 1.34509	valid_0's l2: 1.80925<br>

### 1.3479388206163994<br>

##  Result can be improved further. with additional rounds and new features. This result can be used further to  derive inferences on the transaction amounts
